pd/src/consensus/worker.rs

```diff
@@ -247,164 +241,235 @@ impl Worker {
             .height
             .try_into()
             .expect("height should be nonnegative");
-        let epoch = pending_block.set_height(height);
+        let epoch = pending_block.set_height(
+            height,
+            self.state
+                .private_reader()
+                .chain_params_rx()
+                .borrow()
+                .epoch_duration,
+        );
 
         tracing::debug!(?height, ?epoch, end_height = ?epoch.end_height());
 
+        // Find out which validators were slashed in this block
+        let slashed_validators = pending_block
+            .validator_state_changes
```

Comment:
@zbuc is this where this information will be found in the refactored version you're working on? Can you take a look here to see what needs to be adjusted?